PEP 712: Adding a "converter" parameter to dataclasses.field (#3095)
python/peps: pep-0712.rst

PEP: 712
Title: Adding "converter" dataclasses field specifier parameter
Author: Joshua Cannon <joshdcannon@gmail.com>

> Sponsor: Eric V. Smith <eric at trueblade.com>

Reviewer: Yes, confirmed.

> Sponsor: Eric V. Smith <eric at trueblade.com>

Reviewer 2: Thanks!

> Status: Draft

Author: With https://discuss.python.org/t/draft-pep-adding-converter-dataclasses-field-specifier-parameter/23213/24, I couldn't find out how to just edit the title to use it and keep the conversation going.

Should I still open a new one?

> Status: Draft

Author: done https://github.com/python/peps/pull/3120

Type: Standards Track
Content-Type: text/x-rst

> Created: 01-Jan-2023

Author: Am I allowed to target 3.12? The sooner the better.

> Created: 01-Jan-2023

Reviewer 3: It's probably not realistic to target 3.12 at this point, as the feature freeze is in about three weeks. That's not enough time for a PEP discussion, SC decision, and implementation.

  def converter(x) -> tuple: ...
  def converter(x) -> Sequence: ...

  @overload
  def converter(x) -> tuple: ...
  @overload
  def converter(x: int, y: str) -> list: ...
  def converter(x=..., y = ...): ...

Reference Implementation
========================

The `attrs <#attrs-converters>`_ library already includes a ``converter``
parameter containing converter semantics.

CPython support can be seen on a branch: `GitHub <#attrs-converters>`_.

Rejected Ideas
==============

Just adding "converter" to ``dataclass_transform``'s ``field_specifiers``
-------------------------------------------------------------------------

The idea of isolating this addition to ``dataclass_transform`` was briefly
discussed in `Typing-sig <#only-dataclass-transform>`_ where it was suggested
to open this to ``dataclasses``.

Additionally, adding this to ``dataclasses`` ensures anyone can reap the
benefits without requiring additional libraries.

Automatic conversion using the field's type
-------------------------------------------

One idea could be to allow the type of the field specified (e.g. ``str`` or
``int``) to be used as a converter for each argument provided.

> `Pydantic's data conversion <#pydantic-data-conversion>`_ has semantics which

Reviewer 2: ```suggestion
`Pydantic's data conversion <pydantic-data-conversion_>`__ has semantics which
```

* Fix broken link syntax (as above)
* Remove redundant and confusing `#` in link target (as above)

appear to be similar to this approach.

This works well for fairly simple types, but leads to ambiguity in expected
behavior for complex types such as generics. E.g. For ``tuple[int]`` it is
ambiguous if the converter is supposed to simply convert an iterable to a tuple,
or if it is additionally supposed to convert each element type to ``int``.

Converting the default values
-----------------------------

Having the synthesized ``__init__`` also convert the default values (such as

> ``default`` or the return type of ``default_factory``) when the would make the

Reviewer 2: Appears to be an editing mistake. Simplest solution:

```suggestion
``default`` or the return type of ``default_factory``) would make the
```

expected type of these parameters complex for type-checkers, and does not add
significant value.

> References
> ==========
> .. _#typeshed: https://github.com/python/typeshed
> .. _#attrs-converters: https://www.attrs.org/en/21.2.0/examples.html#conversion
> .. _#cpython-branch: https://github.com/thejcannon/cpython/tree/converter
> .. _#only-dataclass-transform: https://mail.python.org/archives/list/typing-sig@python.org/thread/NWZQIINJQZDOCZGO6TGCUP2PNW4PEKNY/
> .. _#pydantic-data-conversion: https://docs.pydantic.dev/usage/models/#data-conversion

Reviewer 2: ```suggestion
References
==========

.. _attrs-converters: https://www.attrs.org/en/21.2.0/examples.html#conversion
.. _cpython-branch: https://github.com/thejcannon/cpython/tree/converter
.. _only-dataclass-transform: https://mail.python.org/archives/list/typing-sig@python.org/thread/NWZQIINJQZDOCZGO6TGCUP2PNW4PEKNY/
.. _pydantic-data-conversion: https://docs.pydantic.dev/usage/models/#data-conversion
```

* Typeshed is not mentioned or referenced anywhere
* Remove redundant and potentially confusing `#` in link target names (which looks like autonumbered named footnotes, as well as a URL fragment id)
* Fix line breaks

relational mapper (ORM) packages such as SQLAlchemy and Django.

A common feature these libraries provide over the standard library
implementation is the ability for the library to convert arguments given at
initialization time into the types expected for each field using a
user-provided conversion function.

Therefore, this PEP adds a ``converter`` parameter to :func:`dataclasses.field`
(along with the requisite changes to :class:`dataclasses.Field` and
:func:`~py3.11:typing.dataclass_transform`) to specify the function to use to
convert the input value for each field to the representation to be stored in
the dataclass.

Motivation
==========

There is no existing, standard way for :mod:`dataclasses` or third-party
dataclass-like libraries to support argument conversion in a type-checkable
way. To work around this limitation, library authors/users are forced to choose
to:

* Opt-in to a custom Mypy plugin. These plugins help Mypy understand the
  conversion semantics, but not other tools.
* Shift conversion responsibility onto the caller of the dataclass
  constructor. This can make constructing certain dataclasses unnecessarily
  verbose and repetitive.

> * Provide a custom ``__init__`` and which declares "wider" parameter types and

Reviewer 3: ```suggestion
* Provide a custom ``__init__`` which declares "wider" parameter types and
```

  converts them when setting the appropriate attribute. This not only duplicates
  the typing annotations between the converter and ``__init__``, but also opts
  the user out of many of the features :mod:`dataclasses` provides.

> * Not rely on, or ignore type-checking.

Reviewer 3: This doesn't make sense to me, as converters would seem to be a runtime feature that dataclasses simply doesn't support. It's not just about typing.

Author: I'm having trouble mentally parsing your comment mapped back onto the code (it doesn't help that GitHub has displayed several lines as context so I'm not sure exactly the spot you're commenting on).

Can you be more specific or put it another way?

Reviewer 3: I would tell you but I'm about to board a flight to PyCon :). I'll respond when I'm in SLC tonight. If you're also at PyCon, happy to talk in person too.

Author: I'll be there!
How are people finding each other at these things?

Reviewer 3: Never mind, got more time before my flight.

Reading this again, I think my confusion is about the fact that the PEP basically has two goals:

- Adding converter support to `dataclasses` in the stdlib
- Adding converter support to the PEP 681 `dataclass_transform` mechanism

I would expect the PEP to primarily be motivated by the first goal, since that's what the PEP title suggests. And if the PEP is about adding a feature to `dataclasses`, the list entry I commented on doesn't make a lot of sense.

But a lot of your PEP reads like it's primarily motivated by the second goal of extending `dataclass_transform`. That's fine, but it means that there is little motivation for adding this new feature to `dataclasses` in the stdlib. So I think my feedback is that you need to motivate better why we couldn't just add this feature to PEP 681 without changing `dataclasses`.

Reviewer 3: As for PyCon, if you're able I'd encourage you to attend the typing summit (Thursday afternoon) and/or the sprints (Monday onwards); I'll be at both. During the conference itself, people often just meet up at whatever time is convenient. You could perhaps submit a lightning talk or create an open space to talk about your upcoming PEP, so you can get feedback and ideas from more people.

Author: Ok I've certainly led you astray then. My motivation is very much for data classes, as I'm not even using dataclass_transform.

Additionally, whether the feature goes in either, type checkers are responsible for checking the values, which is why a large part of the PEP is dedicated to typing.

library means more users are able to opt-in to these benefits without requiring
third-party libraries. Additionally third-party libraries are able to clue
type-checkers into their own conversion semantics through added support in
:func:`~py3.11:typing.dataclass_transform`, meaning users of those libraries
benefit as well.

Specification
=============

New ``converter`` parameter
---------------------------

This specification introduces a new parameter named ``converter`` to the
:func:`dataclasses.field` function. When an ``__init__`` method is synthesized
by ``dataclass``-like semantics, if an argument is provided for the field, the
``dataclass`` object's attribute will be assigned the result of calling the
converter on the provided argument. If no argument is given and the field was
constructed with a default value, the ``dataclass`` object's attribute will be
assigned the result of calling the converter on the provided default.

Adding this parameter also implies the following changes:

* A ``converter`` attribute will be added to :class:`dataclasses.Field`.
* ``converter`` will be added to the field specifier parameters of arguments

>   provided to :func:`~py3.11:typing.dataclass_transform`'s ``field`` parameter.

Author: I would prefer to be explicit, so I'll steal your first suggestion. Thanks!

Example
'''''''

.. code-block:: python

  @dataclasses.dataclass
  class InventoryItem:
      # `converter` as a type
      id: int = dataclasses.field(converter=int)

>       skus: tuple[int, ...] = dataclasses.field(converter=tuple[int])

Reviewer 3: Do you mean `tuple[int, ...]`? This type would mean a one-element tuple.

Reviewer 2: Oops—fixed that one place in that line (and others) but somehow missed the second one. I think we got all the others already.

      # `converter` as a callable
      names: tuple[str, ...] = dataclasses.field(
        converter=lambda names: tuple(map(str.lower, names))
      )

      # The default value is also converted; therefore the following is not a
      # type error.
      stock_image_path: pathlib.PurePosixPath = dataclasses.field(
        converter=pathlib.PurePosixPath, default="assets/unknown.png"
      )

  item1 = InventoryItem("1", [234, 765], ["PYTHON PLUSHIE", "FLUFFY SNAKE"])
  # item1 would have the following values:
  #   id=1
  #   skus=(234, 765)
  #   names=('python plushie', 'fluffy snake')
  #   stock_image_path=pathlib.PurePosixPath("assets/unknown.png")

Impact on typing
----------------

``converter`` arguments are expected to be callable objects which accept a

> unary argument and return a type compatible with the field's annotated type.

Reviewer 3: ```suggestion
single argument and return a type compatible with the field's annotated type.
```

> The callable's unary argument's type is used as the type of the parameter in

Reviewer 3: ```suggestion
The type of the callable's argument is used as the type of the corresponding parameter in
```

Reviewer 2: Hmm—I actually thought I already submitted a very similar suggestion for this line? Maybe it wasn't applied or there was another very similar line that I did spot.

parameter in the synthesized ``__init__`` method.

> Type-narrowing the argument type

Author: Ahh, it's finally clicked how to best represent this in prose. Thanks for the suggestion!

''''''''''''''''''''''''''''''''

For the purpose of deducing the type of the argument in the synthesized
``__init__`` method, the ``converter`` argument's type can be "narrowed" using
the following rules:

* If the ``converter`` is of type ``Any``, it is assumed to be callable with a
  unary ``Any`` typed-argument.

> * All keyword-only parameters can be ignored.

Reviewer 3: Except those without defaults, which would make the converter invalid.

> * ``**kwargs`` with defaults can be ignored.

Reviewer 4: I find this entire section to be very confusing the way it is worded.

I think what you are trying to say is:
A converter must be a callable that accepts a single positional argument, and the parameter type corresponding to this positional argument provides the type of the the synthesized `__init__` parameter associated with the field. In other words, the argument provided for the `converter` parameter must be compatible with `Callable[[T], X]` where `T` is the input type for the converter and `X` is the output type of the converter.

Does that cover it? If so, then I recommend removing all of the additional text about default parameters, `*args`, `**kwargs`, etc. because they are irrelevant and will lead to confusion. All of those rules (and others) fall out naturally from the fact that the type must be compatible with `Callable[[T], X]`.

Author: Yup, that's much easier to type and read.

Sometimes less if more 😅

> * ``*args`` can be ignored if any parameters precede it. Otherwise if ``*args``

Reviewer 4: One big problem I see with this approach is with lambdas. In Python, there's no way to explicitly annotate the parameter types for a lambda, so they must be inferred through bidirectional type inference (i.e. through the context in which the lambda is used). But in the case of the `converter`, there is no available context, so the parameter type will always be `Any` (or `Unknown`, which is pyright's way of designating an "implicit `Any`). I don't see a good solution to this problem other than to recommend that developers never use lambdas for converters if they want proper type checking.

Author: Oh yeah good point. I knew lambdas would be troublesome for converters, but forgot to include it here.

I like calling out the danger of lambdas. Would it be too far to encourage type checkers to discourage lambda usage in this section?

Reviewer 4: I'm not sure what it would mean for "type checkers to discourage lambda usage". I don't think it would be appropriate for a type checker to issue an error or a warning here because it is not a type violation to use a lambda. When the strictest type checking modes are enabled, type checkers could point out that a type cannot be inferred. That may be enough to dissuade some developers from using lambdas here.

Author: Yeah so my question is if that's appropriate to include in a PEP

Reviewer 4: I think it's sufficient to point out the issue with lambda parameter inference. Different type checkers have their own mechanisms for reporting cases where types cannot be inferred (typically only when stricter type checking modes are enabled). I don't think a PEP should prescribe _how_ or _when_ a type checker should report such a condition.

Author: Excellent! Thanks again for your time reading through this 🙏

Leaving default values as-is allows type-checkers and dataclass authors to
expect that the type of the default matches the type of the field. However,
converting default values has two large advantages:

1. Compatibility with attrs. Attrs unconditionally uses the converter to
   convert the default value.

2. Simpler defaults. Allowing the default value to have the same type as
   user-provided values means dataclass authors get the same conveniences as
   their callers.

Automatic conversion using the field's type
-------------------------------------------

One idea could be to allow the type of the field specified (e.g. ``str`` or
``int``) to be used as a converter for each argument provided.
`Pydantic's data conversion <pydantic-data-conversion_>`__ has semantics which
appear to be similar to this approach.

This works well for fairly simple types, but leads to ambiguity in expected
behavior for complex types such as generics. E.g. For ``tuple[int, ...]`` it is
ambiguous if the converter is supposed to simply convert an iterable to a tuple,
or if it is additionally supposed to convert each element type to ``int``.

References

> ==========
> .. _attrs-converters: https://www.attrs.org/en/21.2.0/examples.html#conversion

Reviewer 2: ```suggestion
==========

.. _attrs-converters: https://www.attrs.org/en/21.2.0/examples.html#conversion
```

Small formatting nit—couldn't make it last time since I was viewing the changes in "last-review-delta mode"

  def converter(x: int, y: str) -> list: ...
  def converter(x=..., y = ...): ...

Backward Compatibility
======================

These changes don't introduce any compatibility problems since they

> only introduce new features.

Reviewer 2: Just because something is a new feature doesn't mean it doesn't have the potential to cause any backward compatibility issues. I would check with your sponsor @ericvsmith and/or @JelleZijlstra for more detailed insight on this.

Reviewer 3: I am fine with this wording.

Security Implications
======================

There are no direct security concerns with these changes.

How to Teach This
=================

Documentation and examples explaining the new parameter and behavior will be
added to the relevant sections of the docs site (primarily on
:mod:`dataclasses`) and linked from the *What's New* document.

Reference Implementation
========================

The attrs library `already includes <attrs-converters_>`__ a ``converter``
parameter containing converter semantics.

CPython support is implemented on `a branch in the author's fork <cpython-branch_>`__.

Rejected Ideas
==============

> Just adding "converter" to :func:`~py3.11:~typing.dataclass_transform`'s ``field_specifiers``
> ---------------------------------------------------------------------------------------------

Reviewer 2: ```suggestion
Just adding "converter" to ``typing.dataclass_transform``'s ``field_specifiers``
--------------------------------------------------------------------------------
```

AFAIK, this won't actually work as expected in a section title.

The idea of isolating this addition to

> :func:`~py3.11:~typing.dataclass_transform` was briefly

Reviewer 2: ```suggestion
:func:`~py3.11:typing.dataclass_transform` was briefly
```

Reviewer 2: This doesn't appear to actually have been resolved, as the change was not applied and the error is still present.

Just to note, as I've advised previously, just directly applying the suggestions rather than manually re-implementing them locally, committing and manually resolving each of them would avoid omissions like these (as well as extra work rechecking them).

Reviewer 2: That's usually a good strategy, yeah—one I've particularly found is helpful when it comes to things like code snippits from Stack Overflow. In this case, most of my suggestions (as opposed to comments) here are reST/Sphinx tweaks, formatting fixes, typo corrections and other technicalities that are less things you did wrong or should have done better from the beginning, and more just routine editorial adjustments that are more the realm of a specialist in this particular domain (i.e. a PEP editor like myself).

While I certainly try to explain all the changes, I wouldn't want each PEP author feeling like they have to spend much of their valuable time and effort learning all the ins and outs of that particular role themselves (which will likely never pay for itself in terms of the total time spent between us), as opposed to just doing the minimum on their end to apply the changes and focus on what they do best—writing a great PEP!

That being said, you're the author so its entirely up to you how you want to apply the changes; if you really want to spend the time then no pressure :)

`discussed on Typing-SIG <only-dataclass-transform_>`__ where it was suggested
to broaden this to :mod:`dataclasses` more generally.

Additionally, adding this to :mod:`dataclasses` ensures anyone can reap the
benefits without requiring additional libraries.

Automatic conversion using the field's type
-------------------------------------------

One idea could be to allow the type of the field specified (e.g. ``str`` or
``int``) to be used as a converter for each argument provided.
`Pydantic's data conversion <pydantic-data-conversion_>`__ has semantics which
appear to be similar to this approach.

This works well for fairly simple types, but leads to ambiguity in expected

> behavior for complex types such as generics. E.g. For ``tuple[int]`` it is

Reviewer 2: ```suggestion
behavior for complex types such as generics. E.g. For ``tuple[int, ...]`` it is
```

Missed this before, sorry—presumably this is a tuple of unspecified length

One idea could be to allow the type of the field specified (e.g. ``str`` or
``int``) to be used as a converter for each argument provided.
`Pydantic's data conversion <#pydantic-data-conversion>`_ has semantics which
`Pydantic's data conversion <pydantic-data-conversion_>`__ has semantics which
appear to be similar to this approach.

This works well for fairly simple types, but leads to ambiguity in expected
behavior for complex types such as generics. E.g. For ``tuple[int]`` it is
ambiguous if the converter is supposed to simply convert an iterable to a tuple,
or if it is additionally supposed to convert each element type to ``int``.

Converting the default values
-----------------------------

Having the synthesized ``__init__`` also convert the default values (such as
``default`` or the return type of ``default_factory``) when the would make the
expected type of these parameters complex for type-checkers, and does not add

> significant value.

Reviewer 2: You might want to add the inverse of this, not converting the default values, as a rejected idea with an appropriate explanation particularly given you proposed it initially.

Author: That is a fantastic idea!
